### DISCUSSION TITLE
Allow for templating of pod-template.yaml at service start time or pod creation

### DISCUSSION THREAD

**Discussion by andscoop** [apache/airflow#27650] [General]
### Description

I've been researching how one can template out the pod-template.yaml.

Airflow provides a few control points for this but none of them seem quite right for our use case
- Helm charts: our k8s environment does not support helm
- pod overrides: called for every pod creation, requires existing defaults and extends some fields instead of overwriting

We're currently "templating" via `sed` find and replace at service start time. This works fine, but it is not as elegant as I think it could be.

### Use case/motivation

Not all kubernetes environments support helm. Additionally, since the K8sExecutor is responsible for pod creation, it should have the ability to render templates closer to pod creation. 

This would allow for a single jinja templated `pod-template.yaml` file to be reused across multiple pipeline clusters or in multi-cluster deployments environments. 

If I am missing something, please point me in the right direction.

Thanks!

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
You have not explained "WHEN" the pod_template would change. From what I understand, you want to do it at service deployment time. And seems this is what you want to do - either on service deployment time or based on some manually triggered  event (which would be some kind of event reconfiguring the whole cluster). What you really need is a solution for your deployment not "airflow feature".

What I think you shuld do is to simply use whatever k8s already provides - you are really talking about config map that will be mounted as your pod_template_file:  https://kubernetes.io/docs/concepts/storage/volumes/#configmap

You can mount this config map as a volume/folder available for all pods that need them and then you can use absolutely standard k8s mechanism to apply any changes to it (kubectl apply) including pre-processing it via jinja as you wish (you do not need sed, you can simply use jinja2 CLI for that https://pypi.org/project/jinja2-cli/

Converting it into discussion, but I suggest you try this first and see if you can make it works.

**Comment by elsayed91** [comment]
Hello, sorry to hijack, not entirely sure if this is the same exact issue, and maybe I am just plain out wrong, but this seems mildly related so here goes.

when using KubernetesPodOperator with params, and then using these params inside the yaml like so
```yaml
  containers:
    - name: transferpod
      image: "{{ params.image }}"
````
the content is not substituted. despite pod_template_file being a templated field, I guess it is templated for the string input not the body content? but that does not feel right.

3rd party addons like [Airflow Kubernetes Job Operator](https://pypi.org/project/airflow-kube-job-operator/#description) or [Airflow KubernetesJobOperator](https://github.com/LamaAni/KubernetesJobOperator) do offer this
and even the [SparkKubernetesOperator](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/_api/airflow/providers/cncf/kubernetes/operators/spark_kubernetes/index.html) work similarly like so
```yaml
        t2 = SparkKubernetesOperator(
            task_id="spark_full_refresh",
            namespace=SPARK_NAMESPACE,
            application_file=f"spark.yaml",
            params=spark_job_params,
        )
```
I currently have to either pre-process the templates as s step, or have to remove sections from the yaml and supplement them in the operator (removing image from yaml and adding image in the operator) which works but it does not feel the best. 

Since this is implemented for KubernetesSparkOperator, why is it not implemented in KubernetesPodOperator?

**Reply by potiuk** [reply]
Ok. I looked at the code and it's rather simple - the reason why it does not suport this pattern is that it is not implemented. I am not sure it was intentional or not. It is pretty normal (for people used to Helm) to have K8S resources defined with JINJA templates, so I would also expect this should be possible. But It seems it is not done:

From what I see the current code looks like this:

```
        if self.pod_template_file:
            self.log.debug("Pod template file found, will parse for base pod")
            pod_template = pod_generator.PodGenerator.deserialize_model_file(self.pod_template_file)
            if self.full_pod_spec:
                pod_template = PodGenerator.reconcile_pods(pod_template, self.full_pod_spec)
        elif self.full_pod_spec:
            pod_template = self.full_pod_spec
        else:
            pod_template = k8s.V1Pod(metadata=k8s.V1ObjectMeta())
```

And deserialize_model_file does not process the file with JINJA/our context. But I think it **could** and it seems rather reasonable.

@dstandish @dimberman @ephraimbuddy - is there a particular reason we do not do JINJA processing for `pod_template_file` I am not aware of ? I know it is not processed automatically because it does not have templated_ext defined (and then it would work a bit differently because it would have accepted not only a file but also the template content). But I see no particular reasons why we should not process it via JINJA (other than back-compat).

**Comment by BayanAzima** [comment]
I would like this feature too. I believe it was working in airflow 2.2 with the now deprecated other KPO. Thanks for all the great work all.
